### PR TITLE
Support for Xcode 10.2 and Swift 5; addition of a universal target

### PR DESCRIPTION
**Pull request by bmalbuck**
* Added SwiftChartsUniversal target that builds a fat binary with compile time arch stripping.
* Migrate to Swift 5 while fixing deprecation warnings

### CODE REVIEW

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift [hunk 1]
```diff
@@ -88,7 +88,8 @@ open class ChartGroupedBarsLayer<T: ChartBarModel, U: ChartPointViewBar>: ChartC
         let totalWidth = CGFloat(group.bars.count) * barWidth + ((barSpacing ?? 0) * (maxBarCountInGroup - 1))
         let groupCenter = screenLocCalculator(group.constant.scalar)
         let origin = groupCenter - totalWidth / 2
-        return origin + CGFloat(index) * (barWidth + (barSpacing ?? 0)) + barWidth / 2
+        let barWidthAddition = (barWidth + (barSpacing ?? 0)) + barWidth / 2
+        return origin + CGFloat(index) * barWidthAddition
```
[the remark this answers is not in the document]
Comment (bmalbuck) [reply SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: Right. I did make a mental note on the order of operation. I went ahead and used your suggestion. I will update the PR soon.

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift [hunk 2]
```diff
@@ -111,7 +112,8 @@ open class ChartGroupedBarsLayer<T: ChartBarModel, U: ChartPointViewBar>: ChartC
             max(maxCount, CGFloat(group.bars.count))
         }
         
-        let barWidth = self.barWidth ?? (((groupAvailableLength - ((barSpacing ?? 0) * (maxBarCountInGroup - 1))) / CGFloat(maxBarCountInGroup)))
+        let calculatedBarWidth = (((groupAvailableLength - ((barSpacing ?? 0) * (maxBarCountInGroup - 1))) / CGFloat(maxBarCountInGroup)))
```
[the remark this answers is not in the document]
Comment (bmalbuck) [reply SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: Good suggestion.

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift [hunk 3]
```diff
@@ -88,8 +88,8 @@ open class ChartGroupedBarsLayer<T: ChartBarModel, U: ChartPointViewBar>: ChartC
         let totalWidth = CGFloat(group.bars.count) * barWidth + ((barSpacing ?? 0) * (maxBarCountInGroup - 1))
         let groupCenter = screenLocCalculator(group.constant.scalar)
         let origin = groupCenter - totalWidth / 2
-        let barWidthAddition = (barWidth + (barSpacing ?? 0)) + barWidth / 2
-        return origin + CGFloat(index) * barWidthAddition
+        let calculatedWidthWithSpacing = CGFloat(index) * (barWidth + (barSpacing ?? 0))
```
Comment (i-schuetz) [inline SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: I'd call this something like internalBarOriginScreenLoc. It refers to the position of the bar in the screen, not width.

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift [hunk 4]
```diff
@@ -112,8 +112,13 @@ open class ChartGroupedBarsLayer<T: ChartBarModel, U: ChartPointViewBar>: ChartC
...
         
-        let calculatedBarWidth = (((groupAvailableLength - ((barSpacing ?? 0) * (maxBarCountInGroup - 1))) / CGFloat(maxBarCountInGroup)))
-        let barWidth = self.barWidth ?? calculatedBarWidth
+        let barWidth: CGFloat
+        if let selfBarWidth = self.barWidth {
+            barWidth = selfBarWidth
+        } else {
+            let calculatedWidth = (groupAvailableLength - ((barSpacing ?? 0) * (maxBarCountInGroup - 1)))
+            barWidth = calculatedWidth / CGFloat(maxBarCountInGroup)
+        }
```
Comment (i-schuetz) [inline SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: What about this?
```swift
func calculateBarWidth() -> CGFloat {
    return (((groupAvailableLength - ((barSpacing ?? 0) * (maxBarCountInGroup - 1))) / CGFloat(maxBarCountInGroup)));
}

let barWidth = self.barWidth ?? calculateBarWidth()
```

File: SwiftCharts/Chart.swift [hunk 5]
```diff
@@ -788,11 +789,15 @@ open class ChartView: UIView, UIGestureRecognizerDelegate {
...
-        case .failed: break;
+        case .cancelled:
+            fallthrough
+        case .failed:
+            fallthrough
         case .possible:
 //            sender.state = UIGestureRecognizerState.Changed
-            break;
+            fallthrough
+        default:
```
Comment (i-schuetz) [inline SwiftCharts/Chart.swift]: Let's make this `@unknown default:` such that the compiler emits a warning.

File: SwiftCharts/Layers/ChartPointsTouchHighlightLayer.swift [hunk 6]
```diff
@@ -102,6 +99,8 @@ open class ChartPointsTouchHighlightLayer<T: ChartPoint, U: UIView>: ChartPoints
                     }
                 )
             }
+        default:
```
Comment (i-schuetz) [inline SwiftCharts/Layers/ChartPointsTouchHighlightLayer.swift]: Let's make this `@unknown default:` such that the compiler emits a warning.

File: SwiftCharts/Chart.swift [hunk 7]
```diff
@@ -788,11 +789,15 @@ open class ChartView: UIView, UIGestureRecognizerDelegate {
             
             chart.onPanEnd()
             
-        case .cancelled: break;
-        case .failed: break;
+        case .cancelled:
+            fallthrough
+        case .failed:
+            fallthrough
```
Comment (i-schuetz) [inline SwiftCharts/Chart.swift]: These 2 can be collapsed `case .cancelled, .failed:` (improvement to my code)

**Comment by i-schuetz** [comment]
Thanks @bmalbuck! Submitted a few comments, other than that it looks good. Thanks also @AgapovOne for helping to review!

File: SwiftCharts/Chart.swift (hunk 5 above)
Comment (bmalbuck) [reply SwiftCharts/Chart.swift]: I left in the .possible case due to the commented code. I'ff let you decide how to structure this switch statement going forward.

**Comment by bmalbuck** [comment]
@i-schuetz I applied your suggested changes with one comment.

**Comment by i-schuetz** [comment]
Thanks @bmalbuck and please next time update your branch with rebase instead of merge, to keep the history clean.

**Comment by i-schuetz** [comment]
Also fixed the remaining warnings in the library project.

**Comment by AgapovOne** [comment]
@i-schuetz Why did you bumped minimum iOS version to 10.0?
Library had 8.0 support, that's great. Why you've used 10.0 now?

**Comment by i-schuetz** [comment]
? I didn't. Didn't notice this in the review.

**Comment by AgapovOne** [comment]
Sorry, it wasn't you. Checked merge in whole history, misjudged.
[This](https://github.com/i-schuetz/SwiftCharts/pull/384/commits/9196caa2f13a636be58b7a7609a308d42630cd08) commit with universal target sets minimum target to 10.0. Downloaded last version, tried to reset to 8, successfully built a framework.

I think we should return it to 8.0.

@bmalbuck why did you change to 10.0?

**Comment by i-schuetz** [comment]
Already changed it back and released a new version (0.6.5)!

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift (hunk 3 above)
Comment (SreekanthGudisi) [reply SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: Do change on **calculateConstantScreenLoc** function

    func calculateConstantScreenLoc(_ screenLocCalculator: (Double) -> CGFloat, index: Int, group: ChartPointsBarGroup<T>, barWidth: CGFloat, maxBarCountInGroup: CGFloat) -> CGFloat {
        let totalWidth = CGFloat(group.bars.count) * barWidth + ((barSpacing ?? 0) * (maxBarCountInGroup - 1))
        let groupCenter = screenLocCalculator(group.constant.scalar)
        let origin = groupCenter - totalWidth / 2
        let calculatedWidthWithSpacing = CGFloat(index) * (barWidth + (barSpacing ?? 0))
        return origin + calculatedWidthWithSpacing + barWidth / 2
    }

File: SwiftCharts/Layers/ChartGroupedBarsLayer.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (SreekanthGudisi) [reply SwiftCharts/Layers/ChartGroupedBarsLayer.swift]: Do change on **calculateConstantScreenLoc** function

func calculateConstantScreenLoc(_ screenLocCalculator: (Double) -> CGFloat, index: Int, group: ChartPointsBarGroup<T>, barWidth: CGFloat, maxBarCountInGroup: CGFloat) -> CGFloat {
    let totalWidth = CGFloat(group.bars.count) * barWidth + ((barSpacing ?? 0) * (maxBarCountInGroup - 1))
    let groupCenter = screenLocCalculator(group.constant.scalar)
    let origin = groupCenter - totalWidth / 2
    let calculatedWidthWithSpacing = CGFloat(index) * (barWidth + (barSpacing ?? 0))
    return origin + calculatedWidthWithSpacing + barWidth / 2
}

**Pull request merged**